### DISCUSSION TITLE
New floating-vue package: changes and migration roadmap

### DISCUSSION THREAD

**Discussion by Akryum** [Akryum/v-tooltip#603] [General]
<p align="center">
  <img src="https://user-images.githubusercontent.com/2798204/103449010-0196b900-4ca2-11eb-85b1-33efaedb3b1d.png">
</p>

<br>

# **v-tooltip has been renamed to floating-vue and now uses [floating-ui](floating-ui.com/) instead of popperjs.**

The package has been renamed from `v-tooltip` to `floating-vue`. You can find the new changelog for both v1 (for Vue 2) and v2 (for Vue 3) [here](https://github.com/Akryum/floating-vue/blob/main/CHANGELOG.md). The changelog is currently not pushed to GitHub Releases.

New versions:

|Old|New|Target
|---|---|---|
|v-tooltip v3|floating-vue v1|Vue 2|
|v-tooltip v4|floating-vue v2|Vue 3|

---

*Original post:*

Version 3 will be a big release. Apart from all the refactoring and code cleaning that went into it, changes and migration paths need to be documented. :scroll: 

I hope you will like the new version (despite all the breaking changes) and that it will be you ultimate tooltips and dropdowns library!

### Notes

- **Please note that V3 is currently in alpha, which means the API (and therefore this list) is unstable and may change at anytime.**
- ~Vue 3 support will be implemented in V4 which will be released in parallel to V3 so you can choose the correct version depending on you Vue version.~ It's [available now](https://github.com/Akryum/v-tooltip/tree/vue3/packages/docs/src)! This will prevent me from polluting the code with conditions and hacks to make it work on both versions simultaneously.
- An entirely new documentation website is planned with a full migration guide ([wip here](https://v-tooltip.netlify.app/))
- Bonus: as of the current alpha, 33% lighter than V2! :ok_hand: 

Here is a preview of the most visible implemented changes (checkboxes are for docs):

# Breaking changes

[Migration guide](https://v-tooltip.netlify.app/migration/migration-from-v2.html)

- [x] Now uses popperjs 2 (`@popperjs/core`)
- [x] The global configuration has been entirely remade and streamlined, now features themes
- [x] Modifiers API has changed, now an array of modifiers such as `{ name: 'overflow', enabled: true }`
- [x] Modifiers should be set with the `modifiers` prop
- [x] `trigger` renamed to `triggers` and now must be an array (use empty array instead of `'manual'`)
- [x] `boundariesElement` prop renamed to `boundary`
- [x] `open` prop renamed to `shown`
- [x] `openGroup` prop renamed to `showGroup`
- [x] Default dropdown component is now `<VDropdown>` instead of `<v-popover>`
- [x] The slot `popover` was renamed to `popper`
- [x] `isOpen` slot prop renamed to `shown`
- [x] `offset` prop must now be an array: `[skidding, distance]`
- [x] CSS classes have been rewritten with the `v-popper` prefix and in a BEM-like fashion to help preventing conflicts
- [x] Removed props: `popoverClass`, `popoverBaseClass`, `popoverWrapperClass`, `popoverArrowClass`, `popoverInnerClass`, `openClass`
- [x] The `v-close-popover` directive has been renamed to `v-close-popper`
- [x] The component HTML structure has been simplified
- [x] As a result events listeners (most notably focus) and `aria-describedby` attribute are now applied on your elements directly instead of an internal `div`
- [x] The component root element now has `width: max-content` by default to use the content size instead of the page width
- [x] The `v-tooltip` directive now uses the `TooltipDirective` component (using `Popper` component internally), which means that there are no longer separate APIs between the components and the directive (the directive has a few additional props though) and you should now be using the components props on the directive object when needed
- [x] Removed `v-tooltip` directive options: `classes`, `targetClasses`, `template`, `arrowSelector`, `innerSelector`, `hideOnTargetClick` (replaced by `hideTriggers`), `loadingClass`

# New

- [x] New powerful & flexible theme system ([example](https://github.com/Akryum/v-tooltip/discussions/604))
- [x] Themes can extend another theme
- [x] Default styles (an import away)
- [x] `strategy` prop
- [x] `modifiers` prop
- [x] `showTriggers` prop
- [x] `hideTriggers` prop
- [x] `popperTriggers` prop
- [x] `popperShowTriggers` prop
- [x] `popperHideTriggers` prop
- [x] `instantMove` prop
- [x] New `VDropdown` component with `dropdown` default theme
- [x] New `VTooltip` component with the `tooltip` default theme
- [x] New `VMenu` component with the `menu` default theme extending the `dropdown` theme, with hover trigger on both the target and the popper itself
- [x] Fully customizable do-it-yourself poppers with `<Popper>`, `<PopperContent>`, `<PopperWrapper>` components and slots

**Comment by sjblasko** [comment]
Out of curiosity, how is the version 3 alpha going?  We are in the process of migrating a legacy .NET application to utilize a Vue 3 front end and would love to utilize this component.  I know you had mentioned that version 4 would be the release that would support Vue 3 and we were just wondering if you or the community had a timeline for the version 3 and 4 releases.  Thanks for everything you do!

**Reply by Akryum** [reply]
Very soon!

**Reply by sjblasko** [reply]
@Akryum, any news?  We are looking to utilize the Tooltips and the VDropdown very soon and the converted version of this project only has working Tooltips (https://github.com/paulshen/v-tooltip).  The VPopover is completely broken in the Vue3 project.

**Reply by Akryum** [reply]
@sjblasko It's now available with `yarn add v-tooltip@vue3`

**Comment by Akryum** [comment]
The Vue 3-compatible version is now available! You can install it using the `4.0.0-alpha.1` version or using the `vue3` tag:

```
yarn add v-tooltip@vue3
```

**Reply by jaulz** [reply]
Thanks for your work 😊  As a vue2 user, should I better use version 3 or 4?

**Reply by Akryum** [reply]
V3 => Vue 2
V4 => Vue 3

**Comment by eugenevk** [comment]
Can I use npm to install the Vue3 version? Tried
```
npm install --save v-tooltip@vue3
```
but that doesn't seem to work.

**Reply by Akryum** [reply]
What isn't working?

I can see the npm tag is correctly published:
![image](https://user-images.githubusercontent.com/2798204/116395145-06b8de80-a824-11eb-8362-07cef0b007ee.png)

**Reply by eugenevk** [reply]
![image](https://user-images.githubusercontent.com/19785727/119327980-ded26480-bc83-11eb-96fd-4cca6b5d654b.png)

**Reply by eugenevk** [reply]
![image](https://user-images.githubusercontent.com/19785727/119328111-0d503f80-bc84-11eb-83d1-faeb40b806bb.png)

**Reply by Akryum** [reply]
Update `vue`

**Reply by eugenevk** [reply]
@Akryum Hi Guillaume, I managed to update vue and now the install is fine. However, I added styling (exactly your css example) in my project's global CSS file and the popup doesn't get formatted:
![image](https://user-images.githubusercontent.com/19785727/119461705-41893600-bd40-11eb-99b6-66b277e3ddbd.png)
Any idea what could be wrong?
I have this in my global CSS file (that gets loaded in my index.html):
```
.tooltip {
    display: block !important;
    z-index: 10000;
}
.tooltip .tooltip-inner {
    background: black;
    color: white;
    border-radius: 16px;
    padding: 5px 10px 4px;
}
.tooltip .tooltip-arrow {
    width: 0;
    height: 0;
    border-style: solid;
    position: absolute;
    margin: 5px;
    border-color: black;
}
.tooltip[x-placement^="top"] {
    margin-bottom: 5px;
}
.tooltip[x-placement^="top"] .tooltip-arrow {
    border-width: 5px 5px 0 5px;
    border-left-color: transparent !important;
    border-right-color: transparent !important;
    border-bottom-color: transparent !important;
    bottom: -5px;
    left: calc(50% - 5px);
    margin-top: 0;
    margin-bottom: 0;
}
.tooltip[x-placement^="bottom"] {
    margin-top: 5px;
}
.tooltip[x-placement^="bottom"] .tooltip-arrow {
    border-width: 0 5px 5px 5px;
    border-left-color: transparent !important;
    border-right-color: transparent !important;
    border-top-color: transparent !important;
    top: -5px;
    left: calc(50% - 5px);
    margin-top: 0;
    margin-bottom: 0;
}
.tooltip[x-placement^="right"] {
    margin-left: 5px;
}
.tooltip[x-placement^="right"] .tooltip-arrow {
    border-width: 5px 5px 5px 0;
    border-left-color: transparent !important;
    border-top-color: transparent !important;
    border-bottom-color: transparent !important;
    left: -5px;
    top: calc(50% - 5px);
    margin-left: 0;
    margin-right: 0;
}
.tooltip[x-placement^="left"] {
    margin-right: 5px;
}
.tooltip[x-placement^="left"] .tooltip-arrow {
    border-width: 5px 0 5px 5px;
    border-top-color: transparent !important;
    border-right-color: transparent !important;
    border-bottom-color: transparent !important;
    right: -5px;
    top: calc(50% - 5px);
    margin-left: 0;
    margin-right: 0;
}
.tooltip[aria-hidden='true'] {
    visibility: hidden;
    opacity: 0;
    transition: opacity .15s, visibility .15s;
}
.tooltip[aria-hidden='false'] {
    visibility: visible;
    opacity: 1;
    transition: opacity .15s;
}
```
and this is my main.js:
```
import { createApp } from 'vue'
import App from './App.vue'
import router from './router'
import VTooltip from 'v-tooltip'

const app = createApp(App)

app.use(router).mount('#app')
app.use(VTooltip)
```
I am also using [TailwindCSS](https://tailwindcss.com/) in my project, but that shouldn't be a problem I assume, unless it interferes of course.

**Reply by eugenevk** [reply]
I think it's the same problem as what @valanto describes below. It looks like she is also using TailwindCSS.

**Reply by valanto** [reply]
@eugenevk I am using tailwind also but haven't tried yet to add a custom style it after getting it install. I will let you know once I try it

**Reply by eugenevk** [reply]
@Akryum I also noticed this warning/error:
![image](https://user-images.githubusercontent.com/19785727/119468351-a6e02580-bd46-11eb-8e8e-eadf973f5321.png)

**Reply by eugenevk** [reply]
> @eugenevk I am using tailwind also but haven't tried yet to add a custom style it after getting it install. I will let you know once I try it

@valanto Have you been able to get your tooltips styled in any way? Or are they not showing any styling as well, like mine?

**Reply by valanto** [reply]
@eugenevk They are working for me now with default styling
![working-v-tooltip](https://user-images.githubusercontent.com/1709204/119468932-96807880-bd4f-11eb-861a-1de664bb04e8.gif)

I think what you are missing is in main.js to import the default style of the tooltip
`import "v-tooltip/dist/v-tooltip.css";`

See more information here: https://v-tooltip.netlify.app/guide/#quick-start

**Reply by eugenevk** [reply]
Hero! You made my day @valanto ! Solved. Case closed @Akryum

**Reply by eugenevk** [reply]
Only interesting additional thing for me would be if TailwindCSS can be used to specify specific classes

**Comment by valanto** [comment]
@Akryum thanks for a great component! Just started a new vue3 project and added v-tooltip as mentioned above but even thought the text shows up I cannot set any global options and it appears unformatted and with a long delay to hide. Is there a way to fix this? What I have in my vue setup (Even without passing options I still get the same behaviour/display):

```ts
import VTooltip from "v-tooltip";
...
app.use(VTooltip, {
  defaultDelay: { show: 500, hide: 500 },
  classes: "bg-neutral-700 opacity-75 rounded-md text-white",
});
...
```

This is what I see: 
![v-tooltip-vue3](https://user-images.githubusercontent.com/1709204/119327237-6d96af80-bc8b-11eb-826c-9fb4f44879d0.gif)

**Reply by Akryum** [reply]
Those are old v2 options, so they won't work. As described in the first post, the new docs is [here](https://v-tooltip.netlify.app/).

**Reply by Akryum** [reply]
> it appears unformatted and with a long delay to hide

In your recording, it appears to have no CSS and simply gets destroyed after a few seconds.

**Reply by valanto** [reply]
Thanks for getting back to me. New instructions seem to work but just had  to use  `yarn add v-tooltip@vue3` rather than  `yarn add v-tooltip@next`

**Reply by SenDen11** [reply]
I've got the same issue with hide delay. The tooltip gets disappeared after about 5 sec (I guess its because of default value of `disposeTimeout`).
Seems like `$_computeDelay` function in Popper.js returns `NaN` as the second parameter for SetTimeout

The version of package is 4.0.0-beta.17
![image](https://user-images.githubusercontent.com/50140751/165553077-a4d97b49-4913-4cc7-8fc1-7dbc5603994c.png)

**Comment by bosquig** [comment]
Hi @Akryum 

I think there might be a bug, or I don't get how I'm supposed to make it work with the new version (migrated from vue 2 to vue 3, so I had to migrate v-tooltip as well).

Env for testing: vite default template with vue 3 with compat build (compatConfig{MODE: 3}), tailwindcss and v-tooltip@vue3 (installed in vue with app.use(VTooltip))

```
<template>
  <div v-tooltip="{
      content: 'Mais si je suis là, regarde !',
      shown: isOpen,
      triggers: [''],
      placement: 'bottom',
    }"  @mouseleave="mouseleave" @mouseenter="mouseenter" class="max-w-16 whitespace-no-wrap overflow-ellipsis overflow-hidden">
    Il s'agit d'une fâcheuse histoire au sujet d'un texte qui n'avait plus de tooltip :(
  </div>
<template>
<script>
export default {
  data() {
    return {
      isOpen: false,
    }
  },
  methods: {
    mouseenter() {
      if (this.$el.offsetWidth < this.$el.scrollWidth) {
        this.isOpen = true;
      }
    },
    mouseleave() {
      this.isOpen = false;
    }
  },
}
</script>
```

throws the following error on mounting : 

Uncaught TypeError: Cannot read property 'tooltip' of undefined
    at bind (v-tooltip.js:99)

Sneak peak at this line:
```value.shown ? tooltipApp.$refs.tooltip.show() : tooltipApp.$refs.tooltip.hide()```
where tooltipApp.$refs is undefined (tooltipApp is using vue version 3.1.0-beta.5)

Note that if I replace  "shown: isOpen,  triggers: ['']" with "triggers: ['hover']" it works (but this is not what I want to do)

Thanks in advance !

EDIT: I managed to get the result I wanted with triggers: ['hover'] and the fact that nothing shows up if the tooltip content is an empty string.

v-tooltip={content: tooltipText, [...]}
data() {...tooltipText: ''}
mounted() { if(this.$el.offsetWidth < this.$el.scrollWidth) this.tooltipText = 'Mais si je suis là, regarde !' }

seems hacky though.

**Reply by Akryum** [reply]
Try `triggers:[]`

**Reply by bosquig** [reply]
Oh my god, I feel so dumb. The worst part is that I used triggers: [] succesfully in a v-dropdown previously...

Thanks :) !

**Comment by hottabych07** [comment]
![Screenshot_141](https://user-images.githubusercontent.com/32727311/120624244-b23ff900-c48a-11eb-9496-715cb57973eb.png)
Hi, I get an error when importing, who knows how to solve the problem?

**Reply by valanto** [reply]
This is a typescript error but the tooltip should still work. To get rid of the error you need to declare the module by adding a file containing `declare module "v-tooltip";` in your `@types` folder. If you don't have a `@types` folder you can add it under `src` for example and then make sure to mention it in your tsconfig.json:

`"typeRoots": ["node_modules/@types", "src/@types"],`

I hope this helps

**Reply by hottabych07** [reply]
Yes thanks, it works. Now I am trying to open a tooltip depending on isOpen and it throws an error: Cannot read property 'tooltip' of undefined

With "triggers: []", also not working...

This works in the old version. What could have gone wrong?
![Screenshot_142](https://user-images.githubusercontent.com/32727311/120633695-a0fbea00-c494-11eb-96a1-8c19804e51e3.png)

@Akryum help pls

**Reply by Akryum** [reply]
https://v-tooltip.netlify.app/guide/directive.html#manual-trigger-example

**Reply by damianosz** [reply]
@Akryum Why would I get this console warning:
[Vue warn]: Failed to resolve component: v-popover 

When using:
```
<v-popover offset="16" popoverClass="multi-line" trigger="click" :autoHide="false">
    <!-- This will be the popover target (for the events and position) -->
    <button class="tooltip-target b3">Click me</button>

    <!-- This will be the content of the popover -->
    <template v-slot:popover>
        <p>
            Some text
        </p>

        <!-- You can put other components too -->
        <SzInput />
    </template>
</v-popover>
```

**Reply by Akryum** [reply]
There is no `<v-popover>` component anymore. See https://v-tooltip.netlify.app/guide/component.html

**Reply by Akryum** [reply]
You can also use [themes](https://v-tooltip.netlify.app/guide/themes.html#custom-theme) to create something else than dropdown/tooltip/hover menu.

**Comment by whaliendev** [comment]
Hi, @Akryum 
I encountered a problem when using `v-tooltip@vue3`, but I am not sure if it's a bug in `v-tooltip`.

I was migrating a `vue-calendar-heatmap` component into vue3, so I changed the dependency to V4, then the error occured:
![image](https://user-images.githubusercontent.com/43714297/139438843-0bbdd566-668a-4076-9035-88659568dce6.png)

It seems that the bug is in lifecycle hooks, then I did a single-step debugging, and then I found that it threw an error in the `beforeMount` lifecycle( i.e., the `bind` hook in vue2 ).
Specifically, it failed to mount to the target element, which is showd below: 
![image](https://user-images.githubusercontent.com/43714297/139440188-4738b98a-6bbd-4754-808f-f1d695d12414.png)

What's more, I noticed that in the dom tree, the html structure is generated successfully, but the text content is not generated in the dom tree. 
![image](https://user-images.githubusercontent.com/43714297/139440421-ef16dc7b-994c-49f7-a392-fd92641f5d21.png)
Can you figure out what's wrong with my project?

**Comment by Akryum** [comment]
The Migration guide (wip) is now available [here](https://v-tooltip.netlify.app/migration/migration-from-v2.html)!

**Comment by aloiseau17** [comment]
Hello @Akryum is there any news about the v3 release ? Is it still under beta version ?

Thank you :)

**Reply by RihanArfan** [reply]
I am _beyond_ confused as releases show one thing (https://github.com/Akryum/floating-vue/releases), changelog shows another (https://github.com/Akryum/floating-vue/blob/main/CHANGELOG.md) and the documentation another (https://floating-vue.starpad.dev/guide/installation.html#compatibility). I'm hoping @Akryum can shed more light into what the currently developed version is.

**Reply by Akryum** [reply]
The latest versions are floating-vue 1 (for Vue 2) and floating-vue 2 (for Vue 3). The old GH releases are from v-tooltip. So both the CHANGELOG.md (listing versions for both floating-vue 1 and 2) and the docs are correct.
